### PR TITLE
Don't try to hoist/sink {begin,end}_access when the loop has no exits

### LINKED ISSUE #67084: LICM: Could not perform must-sink instruction
**Description**

The compiler crashes with message `LICM: Could not perform must-sink instruction`.

**Steps to reproduce**

```
$ cat loop.swift
var i = 0
while true {
  i += 1
}
$ ~/swift-project-5.8.1/build/Ninja-RelWithDebInfoAssert/swift-linux-x86_64/bin/swiftc loop.swift -O
LICM: Could not perform must-sink instruction
UNREACHABLE executed at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp:489!
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/hiroshi/swift-project-5.8.1/build/Ninja-RelWithDebInfoAssert/swift-linux-x86_64/bin/swift-frontend -frontend -c -primary-file loop.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -color-diagnostics -O -module-name loop -o /tmp/loop-0a5869.o
1.	Swift version 5.8.1-dev (LLVM f0fb631dd1a3a29, Swift 32b31b9dd521ce3)
2.	Compiling with the current language version
3.	While evaluating request ExecuteSILPipelineRequest(Run pipelines { PrepareOptimizationPasses, EarlyModulePasses, HighLevel,Function+EarlyLoopOpt, HighLevel,Module+StackPromote, MidLevel,Function, ClosureSpecialize, LowLevel,Function, LateLoopOpt, SIL Debug Info Generator } on SIL for loop)
4.	While running pass #71 SILFunctionTransform "HighLevelLICM" on SILFunction "@main".

(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737339770816) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737339770816) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737339770816, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff7293476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff72797f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x000055555b9aab11 in llvm::llvm_unreachable_internal (msg=<optimized out>, file=0x55555d0f99f9 "/home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp", line=line@entry=489)
    at /home/hiroshi/swift-project-5.8.1/llvm-project/llvm/lib/Support/ErrorHandling.cpp:209
#6  0x0000555556dcacd3 in (anonymous namespace)::hoistSpecialInstruction (LoopSummary=std::unique_ptr<(anonymous namespace)::LoopNestSummary> = {...}, DT=0x55555f985fe0, LI=0x55555f770010, Special=...)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp:489
#7  (anonymous namespace)::LoopTreeOptimization::optimizeLoop (this=<optimized out>, CurrSummary=...) at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp:1156
#8  (anonymous namespace)::LoopTreeOptimization::optimize (this=<optimized out>) at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp:634
#9  (anonymous namespace)::LICM::run (this=0x55555f986990) at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/LoopTransforms/LICM.cpp:1485
#10 0x00005555569f92fe in swift::SILPassManager::runPassOnFunction (this=this@entry=0x7fffffff9aa0, TransIdx=64, F=<optimized out>)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/PassManager.cpp:582
#11 0x00005555569fa2dc in swift::SILPassManager::runFunctionPasses (this=0x7fffffff9aa0, FromTransIdx=0, ToTransIdx=<optimized out>)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/PassManager.cpp:700
#12 0x00005555569f7588 in swift::SILPassManager::executePassPipelinePlan (this=this@entry=0x7fffffff9aa0, Plan=...)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/PassManager.cpp:824
#13 0x00005555569f753b in swift::ExecuteSILPipelineRequest::evaluate (this=<optimized out>, evaluator=..., desc=...)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/PassManager.cpp:343
#14 0x0000555556a22eea in swift::SimpleRequest<swift::ExecuteSILPipelineRequest, std::tuple<> (swift::SILPipelineExecutionDescriptor), (swift::RequestFlags)1>::callDerived<0ul>(swift::Evaluator&, std::integer_sequence<unsigned long, 0ul>) const (this=0x11782c, evaluator=...) at /home/hiroshi/swift-project-5.8.1/swift/include/swift/AST/SimpleRequest.h:267
#15 swift::SimpleRequest<swift::ExecuteSILPipelineRequest, std::tuple<> (swift::SILPipelineExecutionDescriptor), (swift::RequestFlags)1>::evaluateRequest(swift::ExecuteSILPipelineRequest const&, swift::Evaluator&) (request=..., evaluator=...) at /home/hiroshi/swift-project-5.8.1/swift/include/swift/AST/SimpleRequest.h:290
#16 0x0000555556a059f3 in swift::Evaluator::getResultUncached<swift::ExecuteSILPipelineRequest> (this=0x55555f51e688, request=...)
    at /home/hiroshi/swift-project-5.8.1/swift/include/swift/AST/Evaluator.h:360
#17 0x00005555569f7755 in swift::Evaluator::operator()<swift::ExecuteSILPipelineRequest, (void*)0> (this=0x11782c, this@entry=0x0, request=...)
    at /home/hiroshi/swift-project-5.8.1/swift/include/swift/AST/Evaluator.h:275
#18 swift::executePassPipelinePlan (SM=SM@entry=0x55555f9791d0, plan=..., isMandatory=false, IRMod=IRMod@entry=0x0)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/PassManager.cpp:353
#19 0x0000555556a08aaf in swift::runSILOptimizationPasses (Module=...) at /home/hiroshi/swift-project-5.8.1/swift/lib/SILOptimizer/PassManager/Passes.cpp:116
#20 0x00005555564d29cf in performSILOptimizations (Invocation=..., SM=0x55555f9791d0) at /home/hiroshi/swift-project-5.8.1/swift/lib/Frontend/Frontend.cpp:1358
#21 swift::CompilerInstance::performSILProcessing (this=0x55555f51c120, silModule=0x55555f9791d0) at /home/hiroshi/swift-project-5.8.1/swift/lib/Frontend/Frontend.cpp:1392
#22 0x00005555562af3e2 in performCompileStepsPostSILGen (Instance=..., SM=std::unique_ptr<swift::SILModule> = {...}, MSF=..., PSPs=..., ReturnValue=@0x7fffffffae34: 0, observer=observer@entry=0x0)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:1717
#23 0x00005555562ae4dd in swift::performCompileStepsPostSema (Instance=..., ReturnValue=@0x7fffffff8c10: 32, observer=0xcccccccd)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:850
#24 0x00005555562c13a4 in llvm::function_ref<bool (swift::CompilerInstance&)>::operator()(swift::CompilerInstance&) const (params=..., this=<optimized out>)
    at /home/hiroshi/swift-project-5.8.1/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:68
#25 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) (Instance=..., observer=<optimized out>, cont=..., 
    runDespiteErrors=false) at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:1241
#26 0x00005555562b1243 in performAction (Instance=..., ReturnValue=<optimized out>, observer=0x0) at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:1368
#27 performCompile (Instance=..., ReturnValue=@0x7fffffffae34: 0, observer=observer@entry=0x0) at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:1409
#28 0x00005555562b0237 in swift::performFrontend (Args=..., Argv0=<optimized out>, MainAddr=<optimized out>, observer=0x6, observer@entry=0x0)
    at /home/hiroshi/swift-project-5.8.1/swift/lib/FrontendTool/FrontendTool.cpp:2323
#29 0x00005555560b8b19 in run_driver (ExecName=..., argv=..., originalArgv=...) at /home/hiroshi/swift-project-5.8.1/swift/lib/DriverTool/driver.cpp:215
#30 swift::mainEntry (argc_=<optimized out>, argv_=0x7fffffffe318) at /home/hiroshi/swift-project-5.8.1/swift/lib/DriverTool/driver.cpp:418
#31 0x00007ffff727ad90 in __libc_start_call_main (main=main@entry=0x5555560b8520 <main(int, char const**)>, argc=argc@entry=14, argv=argv@entry=0x7fffffffe318) at ../sysdeps/nptl/libc_start_call_main.h:58
#32 0x00007ffff727ae40 in __libc_start_main_impl (main=0x5555560b8520 <main(int, char const**)>, argc=14, argv=0x7fffffffe318, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7fffffffe308) at ../csu/libc-start.c:392
#33 0x00005555560b8285 in _start ()
```

This happens when the IR looks like the following and after the `begin_access` was hoisted, the compiler tries to sink the corresponding `end_access` but failed because there are no exits in the loop in the LICM pass.

```
// main
sil @main : $@convention(c) (Int32, UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>) -> Int32 {
bb0(%0 : $Int32, %1 : $UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>):
  alloc_global @$s4loop1iSivp                     // id: %2
  %3 = global_addr @$s4loop1iSivp : $*Int         // users: %6, %9
  %4 = integer_literal $Builtin.Int64, 0          // user: %5
  %5 = struct $Int (%4 : $Builtin.Int64)          // user: %6
  store %5 to %3 : $*Int                          // id: %6
  %7 = integer_literal $Builtin.Int64, 1          // user: %13
  %8 = integer_literal $Builtin.Int1, -1          // user: %13
  %9 = begin_access [modify] [dynamic] [no_nested_conflict] %3 : $*Int // users: %18, %11, %19
  br bb1                                          // id: %10

bb1:                                              // Preds: bb0 bb1
  %11 = struct_element_addr %9 : $*Int, #Int._value // user: %12
  %12 = load %11 : $*Builtin.Int64                // user: %13
  %13 = builtin "sadd_with_overflow_Int64"(%12 : $Builtin.Int64, %7 : $Builtin.Int64, %8 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %15, %14
  %14 = tuple_extract %13 : $(Builtin.Int64, Builtin.Int1), 0 // user: %17
  %15 = tuple_extract %13 : $(Builtin.Int64, Builtin.Int1), 1 // user: %16
  cond_fail %15 : $Builtin.Int1, "arithmetic overflow" // id: %16
  %17 = struct $Int (%14 : $Builtin.Int64)        // user: %18
  store %17 to %9 : $*Int                         // id: %18
  end_access %9 : $*Int                           // id: %19
  br bb1                                          // id: %20
} // end sil function 'main'

$3 = void
```

**Expected behavior**

The compiler doesn't crash.

This happens in the compiler in version 5.8.1 or the main branch (5.9-dev) when built with assertions enabled across Mac/Linux/Windows.

### PR DESCRIPTION
**Pull request by hjyamauchi**
This fixes compiler crash with message `LICM: Could not perform must-sink instruction`.

Fixes: #67084

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
CC: @gottesmm @atrick @eeckstein

File: test/SILOptimizer/dont_hoist_begin_access_loop_with_no_exit.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// RUN: %target-swiftc_driver -O %s
+
+// Issue URL
```
Comment (compnerd) [inline test/SILOptimizer/dont_hoist_begin_access_loop_with_no_exit.swift:3]: I assume that you meant:

```suggestion
// https://github.com/apple/swift/issues/67084
```

**Review by eeckstein — APPROVED** [review]
lgtm, thanks!

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by hjyamauchi** [comment]
> org.jenkinsci.plugins.workflow.steps.FlowInterruptedException

I think the Linux CI failure is an infra issue. @compnerd, do you mind rerunning the CI?

**Comment by eeckstein** [comment]
@swift-ci test linux

**Pull request merged**